d2/src/main/java/com/linkedin/d2/balancer/servers/AnnouncerUriHashPrefixGenerator.java

```diff
@@ -0,0 +1,54 @@
+/*
+   Copyright (c) 2020 LinkedIn Corp.
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+package com.linkedin.d2.balancer.servers;
+
+import com.linkedin.d2.discovery.stores.zk.ZooKeeperEphemeralStore;
+import com.linkedin.d2.discovery.stores.zk.ZookeeperEphemeralPrefixGenerator;
+
+/**
+ * EphemeralPrefixGenerator helps to create a hash based prefix for Ephemeral Nodes in {@link ZooKeeperEphemeralStore}
+ * The hash generated by EphemeralPrefixGenerator will be based on the URI of the {@link ZooKeeperAnnouncer}
+ * This hash prefixing of EphemeralNodes helps {@link ZooKeeperEphemeralStore} to optimize the reads during markUp/markDown
+ * @author Nizar Mankulangara (nmankulangara@linkedin.com)
+ */
+
+public class AnnouncerUriHashPrefixGenerator implements ZookeeperEphemeralPrefixGenerator
+{
+  private final ZooKeeperAnnouncer _announcer;
+
+  public AnnouncerUriHashPrefixGenerator(ZooKeeperAnnouncer Announcer)
+  {
+    _announcer = Announcer;
+  }
+
+  @Override
+  public String generatePrefix()
+  {
+    return getAnnouncedUriHashCode();
+  }
+
+  public String getAnnouncedUriHashCode()
+  {
+    if (_announcer == null || _announcer.getUri() == null)
+    {
+      return ZooKeeperEphemeralStore.DEFAULT_PREFIX;
+    }
+
+    int hashCode = Math.abs(_announcer.getUri().hashCode());
+    return Integer.toString(hashCode);
+  }
```

Comment:
Still prefer to use the full hostName + others as the prefix. I do not see big benefit using hashing -- you need to generated hash value every time so no CPU savings. More importantly, using the host name as the node name can immediately help me to find the right node when I browse the zk directory. That's very helpful for manual checking!

Reply:
agree on the value of keeping the hostname. Could we append the hostName after the hash code? like "HashCode_HostName_"?

I think the hashCode could save us a few reads on some nodes in the case that many URIs (from multiple applications) are hosted on one host.

Reply:
I mean "HashCode-HostName-" (with hyphens instead of underscores). The filter class could only match the hashCode part, and leave the hostname for manual debugging.

Reply:
Thanks - If we prefer hostname - I am okay with it too. I want to keep it simple and avoid combining both hostname and hashcode as the value of combining is not that significant. Also we may have other repercussions if the size of the znode is too big than the current one.